File: LeetCode/781.Rabbits_in_Forest/Mario/781.Rabbits_in_Forest.java
class Solution {
    public int numRabbits(int[] answers) {
        // {2:2, 3:1} [1, 1, 2]
        // {2:3, 3:1} [1, 1, 1, 2]
        Map<Integer, Integer> counter = new HashMap<>();
        for (int ans : answers) {
            counter.put(ans + 1, counter.getOrDefault(ans + 1, 0) + 1);
        }

        int result = 0;
        for (int key : counter.keySet()) {
            int cnt = counter.get(key);
            int group = cnt / key;
            int remainder = cnt - key * group;
            group += (remainder > 0 ? 1 : 0);
            result += group * key;
        }

        return result;
    }
}

File: LeetCode/781.Rabbits_in_Forest/781.Rabbits_in_Forest.swift
class Solution {
    func numRabbits(_ answers: [Int]) -> Int {
        // cnt = answers[i], cnt + 1
        var counter: [Int: Int] = [:]

        for cnt in answers {
            counter[cnt + 1, default: 0] += 1
        }

        // 1, 1, 1, first two 1 is pair, (3+2-1) / 2 = 2
        // 2 pair, 1 pairs has 2 rabbits, 2 * 2
        var res = 0
        // k is group isze, v : how many answers this ize
        for (k, v) in counter {
            let cnt = (v + k - 1) / k
            res += cnt * k
        }

        return res
    }
}
